gflownet/envs/crystals/crystal.py

```diff
@@ -0,0 +1,488 @@
+            [0]
+            + self.composition.source
+            + self.space_group.source
+            + self.lattice_parameters.source
+        )
+
+        # start and end indices of individual substates
+        self.composition_state_start = 1
+        self.composition_state_end = self.composition_state_start + len(
+            self.composition.source
+        )
+        self.space_group_state_start = self.composition_state_end
+        self.space_group_state_end = self.space_group_state_start + len(
+            self.space_group.source
+        )
+        self.lattice_parameters_state_start = self.space_group_state_end
+        self.lattice_parameters_state_end = self.lattice_parameters_state_start + len(
+            self.lattice_parameters.source
+        )
+
+        # start and end indices of individual submasks
+        self.composition_mask_start = 0
+        self.composition_mask_end = self.composition_mask_start + len(
+            self.composition.action_space
+        )
+        self.space_group_mask_start = self.composition_mask_end
+        self.space_group_mask_end = self.space_group_mask_start + len(
+            self.space_group.action_space
+        )
+        self.lattice_parameters_mask_start = self.space_group_mask_end
+        self.lattice_parameters_mask_end = self.lattice_parameters_mask_start + len(
+            self.lattice_parameters.action_space
+        )
+
+        self.eos = self.lattice_parameters.eos
+        self.stage = Stage.COMPOSITION
+        self.max_action_length = max(
+            max(len(a) for a in self.composition.action_space),
+            max(len(a) for a in self.space_group.action_space),
+            max(len(a) for a in self.lattice_parameters.action_space),
+        )
+        self.done = False
+
+        # Conversions
+        self.state2proxy = self.state2oracle
+        self.statebatch2proxy = self.statebatch2oracle
+        self.statetorch2proxy = self.statetorch2oracle
+
+        super().__init__(**kwargs)
+
+    def _set_lattice_parameters(self):
+        """
+        Sets LatticeParameters conditioned on the lattice system derived from the SpaceGroup.
+        """
+        crystal_system = self.space_group.get_crystal_system()
+
+        if crystal_system is None:
+            raise ValueError(
+                "Cannot set lattice parameters without crystal system determined in the space group."
+            )
+
+        self.lattice_parameters = LatticeParameters(
+            lattice_system=self.space_group.lattice_system,
+            **self.lattice_parameters_kwargs,
+        )
+
+    def _pad_action(self, action: Tuple[int], stage: Stage) -> Tuple[int]:
+        """
+        Pads action such that all actions, regardless of the underlying environment, have
+        the same length. Required due to the fact that action space has to be convertable to
+        a tensor.
+        """
+        return action + (stage.value,) * (self.max_action_length - len(action))
```

Comment:
The multiplication thing is probably fine because it is a tuple but flagging this just in case it could have some unwanted side effects. @vict0rsch ?

Reply:
It's a tuple of ints that is never modified afterwards, I believe it should be find.